### PR TITLE
fix(sfn): handle pagination and ignore errors for expired executions history

### PR DESCRIPTION
**Pull request by pdecat**
This PR updates  the `aws_sfn_state_machine_execution_history` table to implement pagination and ignore errors for expired executions history, [which are only preserved for a maximum of 90 days](https://docs.aws.amazon.com/step-functions/latest/dg/limits-overview.html#service-limits-state-machine-executions).

Without this latter change, one would get errors like this one:
```
Error: operation error SFN: GetExecutionHistory, https response error StatusCode: 400, RequestID: 68dacbfc-af53-42f5-985e-a019c51b679f, ExecutionDoesNotExist: Execution Does Not Exist: 'arn:aw
s:states:eu-west-3:012345678901:execution:sfn-test-012345678901:b1b58243-1276-4195-863a-e83bed3dd603' (SQLSTATE HV000)

```

# Integration test logs
<details>
  <summary>Logs</summary>

```
Add passing integration test logs here
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```
Add example SQL query results here (please include the input queries as well)
```
</details>

### CODE REVIEW

File: aws/table_aws_sfn_state_machine_execution_history.go [hunk 1]
```diff
@@ -346,20 +348,47 @@ func getRowDataForExecutionHistory(ctx context.Context, d *plugin.QueryData, arn
...
+	})
+	// List call
+	for paginator.HasMorePages() {
+		plugin.Logger(ctx).Trace("aws_sfn_state_machine_execution_history.getRowDataForExecutionHistory", "api_call GetExecutionHistory", arn)
+		output, err := paginator.NextPage(ctx)
+		if err != nil {
+			var apiErr smithy.APIError
+			if errors.As(err, &apiErr) {
+				switch apiErr.(type) {
+				case *types.ExecutionDoesNotExist:
```
Comment (cbruno10) [inline aws/table_aws_sfn_state_machine_execution_history.go:379]: Is there a known error code for this error type? If so, could we add it to an `IgnoreConfig` in the `ListConfig`, similar to https://github.com/turbot/steampipe-plugin-aws/blob/d7dac4dcb306a4614b8ef962ecf510a3d7ce8584/aws/table_aws_ec2_instance.go#L29-L31 (this example is in a `GetConfig`, but `ListConfig` supports adding the same thing)?

The `ShouldIgnoreErrorFunc` in the config has a similar effect of returning `nil` if we encounter from the function.

File: aws/table_aws_sfn_state_machine_execution_history.go [hunk 2]
```diff
@@ -346,20 +348,47 @@ func getRowDataForExecutionHistory(ctx context.Context, d *plugin.QueryData, arn
...
+	limit := d.QueryContext.Limit
+	if limit != nil {
+		if *limit < int64(maxLimit) {
+			maxLimit = int32(*limit)
+		}
 	}
 
 	var items []historyInfo
 
-	listHistory, err := svc.GetExecutionHistory(ctx, params)
```
Comment (cbruno10) [inline aws/table_aws_sfn_state_machine_execution_history.go:355]: Does the AWS SDK offer any pagination with `svc.GetExecutionHistory`? Or was this table never iterating through pages before?

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@pdecat Thanks for opening this PR! Can you please see my comments/questions?

File: aws/table_aws_sfn_state_machine_execution_history.go (hunk 2 above)
Comment (pdecat) [reply aws/table_aws_sfn_state_machine_execution_history.go:355]: This table did not implement the required pagination handling before: https://docs.aws.amazon.com/sdk-for-go/api/service/sfn/#SFN.GetExecutionHistory

File: aws/table_aws_sfn_state_machine_execution_history.go (hunk 1 above)
Comment (pdecat) [reply aws/table_aws_sfn_state_machine_execution_history.go:379]: I guess `ExecutionDoesNotExist` should be known: https://github.com/aws/aws-sdk-go-v2/blob/v1.21.2/service/sfn/types/errors.go#L166

Comment (pdecat) [reply aws/table_aws_sfn_state_machine_execution_history.go:379]: I've tried this change:

```
diff --git a/aws/table_aws_sfn_state_machine_execution_history.go b/aws/table_aws_sfn_state_machine_execution_history.go
index c6ae08ce..9384b73a 100644
--- a/aws/table_aws_sfn_state_machine_execution_history.go
+++ b/aws/table_aws_sfn_state_machine_execution_history.go
@@ -2,7 +2,6 @@ package aws

 import (
        "context"
-       "errors"
        "strconv"
        "strings"
        "sync"
@@ -10,7 +9,6 @@ import (
        "github.com/aws/aws-sdk-go-v2/aws"
        "github.com/aws/aws-sdk-go-v2/service/sfn"
        "github.com/aws/aws-sdk-go-v2/service/sfn/types"
-       "github.com/aws/smithy-go"

        sfnv1 "github.com/aws/aws-sdk-go/service/sfn"

@@ -27,6 +25,9 @@ func tableAwsStepFunctionsStateMachineExecutionHistory(_ context.Context) *plugi
                        Hydrate:       listStepFunctionsStateMachineExecutionHistories,
                        Tags:          map[string]string{"service": "states", "action": "ListExecutions"},
                        ParentHydrate: listStepFunctionsStateMachines,
+                       IgnoreConfig: &plugin.IgnoreConfig{
+                               ShouldIgnoreErrorFunc: shouldIgnoreErrors([]string{"ExecutionDoesNotExist"}),
+                       },
                },
                GetMatrixItemFunc: SupportedRegionMatrix(sfnv1.EndpointsID),
                Columns: awsRegionalColumns([]*plugin.Column{
@@ -373,15 +374,6 @@ func getRowDataForExecutionHistory(ctx context.Context, d *plugin.QueryData, arn
                plugin.Logger(ctx).Trace("aws_sfn_state_machine_execution_history.getRowDataForExecutionHistory", "api_call GetExecutionHistory", arn)
                output, err := paginator.NextPage(ctx)
                if err != nil {
-                       var apiErr smithy.APIError
-                       if errors.As(err, &apiErr) {
-                               switch apiErr.(type) {
-                               case *types.ExecutionDoesNotExist:
-                                       // Ignore expired executions for which history is no longer available
-                                       plugin.Logger(ctx).Trace("aws_sfn_state_machine_execution_history.getRowDataForExecutionHistory", "api_error ignore_expired", err)
-                                       return nil, nil
-                               }
-                       }
                        plugin.Logger(ctx).Error("aws_sfn_state_machine_execution_history.getRowDataForExecutionHistory", "api_error", err)
                        return nil, err
                }
```

But it does not seem to work:

```
> with constants (r, s_arn, e_arn) as (
    values (
            'eu-west-3',
            'arn:aws:states:eu-west-3:012345678901:stateMachine:mysfn',
            'arn:aws:states:eu-west-3:012345678901:execution:mysfn:%'
        )
)
select id,
    execution_arn,
    execution_failed_event_details
from aws_sfn_state_machine_execution_history,
    constants
where region = r
    and execution_arn in (
        select execution_arn
        from aws_sfn_state_machine_execution,
            constants
        where region = r
            and state_machine_arn = s_arn
            and execution_arn like e_arn
    )
    and execution_failed_event_details is null;

Error: operation error SFN: GetExecutionHistory, https response error StatusCode: 400, RequestID: 0ce49d57-9f15-4636-8398-8ddaf01af24b, ExecutionDoesNotExist: Execution Does Not Exist: 'arn:aws:states:eu-west-3:012345678901:execution:mysfn:a4fdeb5a-0f8f-4bee-b36b-2ae036665817' (SQLSTATE HV000)

+----+---------------+--------------------------------+
| id | execution_arn | execution_failed_event_details |
+----+---------------+--------------------------------+
+----+---------------+--------------------------------+

```

**Comment by pdecat** [comment]
Not stale

File: aws/table_aws_sfn_state_machine_execution_history.go (hunk 1 above)
Comment (cbruno10) [reply aws/table_aws_sfn_state_machine_execution_history.go:379]: Thanks @pdecat for testing (sorry for the long response time!), I think this is due to a bug where `IgnoreConfig` doesn't work when a `ParentHydrate` is used (https://github.com/turbot/steampipe-plugin-sdk/issues/544), so can you please add a comment mentioning why we need to handle errors here?

Comment (pdecat) [reply aws/table_aws_sfn_state_machine_execution_history.go:379]: Done!

**Review by misraved — APPROVED** [review]

**Pull request merged**